keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Multinomial.java

```diff
@@ -0,0 +1,133 @@
+package io.improbable.keanu.distributions.discrete;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.math3.special.Gamma;
+import org.nd4j.linalg.util.ArrayUtil;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import io.improbable.keanu.distributions.DiscreteDistribution;
+import io.improbable.keanu.tensor.Tensor;
+import io.improbable.keanu.tensor.TensorShapeValidation;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+
+public class Multinomial implements DiscreteDistribution {
+
+    private final IntegerTensor n;
+    private final DoubleTensor p;
+    private final int numCategories;
+
+    public static Multinomial withParameters(IntegerTensor n, DoubleTensor p) {
+        return new Multinomial(n, p);
+    }
+
+    /**
+     * @see <a href="https://en.wikipedia.org/wiki/Multinomial_distribution">Multinomial Distribution</a>
+     * Generalisation of the Binomial distribution to variables with more than 2 possible values
+     *
+     * @param n The number of draws from the variable
+     * @param p The probability of observing each of the k values (which sum to 1)
+     *          p is a Tensor whose first dimension must be of size k
+     */
+    private Multinomial(IntegerTensor n, DoubleTensor p) {
+        Preconditions.checkArgument(
+            p.sum(0).elementwiseEquals(DoubleTensor.ones(n.getShape())).allTrue(),
+            "Probabilities must sum to one"
+        );
+
+        numCategories = p.getShape()[0];
+        for (int category = 0; category < numCategories; category++) {
+            TensorShapeValidation.checkAllShapesMatch(n.getShape(), p.slice(0, category).getShape());
+        }
+        this.n = n;
+        this.p = p;
+    }
+
+    @Override
+    public IntegerTensor sample(int[] shape, KeanuRandom random) {
+        TensorShapeValidation.checkTensorsMatchNonScalarShapeOrAreScalar(shape, n.getShape());
+
+        Tensor.FlattenedView<Integer> nFlattened = n.getFlattenedView();
+        List<Tensor.FlattenedView<Double>> pFlattened = Lists.newArrayList();
```

Comment:
I like what you're doing here and I think it's something people will want to do a lot. Can we pull this into a method on `DoubleTensor`:

```
    @Override
    public List<DoubleTensor> sliceAlongDimension(int dimension, int indexStart, int indexEnd) {
        List<DoubleTensor> slicedTensors = new ArrayList<>();

        for (int i = indexStart; i < indexEnd; i++) {
            slicedTensors.add(slice(dimension, i));
        }

        return slicedTensors;
    }
```

Then you can do:

```
 Tensor.FlattenedView<Integer> nFlattened = n.getFlattenedView();
        List<DoubleTensor> sliced = p.sliceAlongDimension(0, 0, numCategories);

        int length = ArrayUtil.prod(shape);
        int[] samples = new int[0];

        for (int i = 0; i < length; i++) {
            final int j = i;
            List<Double> ps = sliced.stream().map(p -> p.getFlattenedView().getOrScalar(j)).collect(Collectors.toList());
            int[] sample = sample(nFlattened.getOrScalar(i), random, ps.toArray(new Double[0]));
            samples = ArrayUtils.addAll(samples, sample);
        }
```

Reply:
Done - made `sliceAlongDimension` a default method on the `DoubleTensor` interface.